presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedJoinToJoin.java

```diff
@@ -64,27 +77,98 @@ public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Co
 
         PlanNodeDecorrelator planNodeDecorrelator = new PlanNodeDecorrelator(metadata, context.getSymbolAllocator(), context.getLookup());
         Optional<DecorrelatedNode> decorrelatedNodeOptional = planNodeDecorrelator.decorrelateFilters(subquery, correlatedJoinNode.getCorrelation());
+        if (!decorrelatedNodeOptional.isPresent()) {
+            return Result.empty();
+        }
+        DecorrelatedNode decorrelatedSubquery = decorrelatedNodeOptional.get();
 
-        return decorrelatedNodeOptional
-                .map(decorrelatedNode -> {
-                    Expression joinFilter = combineConjuncts(
-                            metadata,
-                            decorrelatedNode.getCorrelatedPredicates().orElse(TRUE_LITERAL),
-                            correlatedJoinNode.getFilter());
-                    return Result.ofPlanNode(new JoinNode(
-                            context.getIdAllocator().getNextId(),
-                            correlatedJoinNode.getType().toJoinNodeType(),
-                            correlatedJoinNode.getInput(),
-                            decorrelatedNode.getNode(),
-                            ImmutableList.of(),
-                            correlatedJoinNode.getOutputSymbols(),
-                            joinFilter.equals(TRUE_LITERAL) ? Optional.empty() : Optional.of(joinFilter),
-                            Optional.empty(),
-                            Optional.empty(),
-                            Optional.empty(),
-                            Optional.empty(),
-                            ImmutableMap.of()));
-                })
-                .orElseGet(Result::empty);
+        // handle INNER and LEFT correlated join
+        if (correlatedJoinNode.getType() == INNER || correlatedJoinNode.getType() == LEFT) {
+            Expression filter = combineConjuncts(
+                    metadata,
+                    decorrelatedSubquery.getCorrelatedPredicates().orElse(TRUE_LITERAL),
+                    correlatedJoinNode.getFilter());
+            return Result.ofPlanNode(rewriteToJoin(
+                    correlatedJoinNode,
+                    correlatedJoinNode.getType().toJoinNodeType(),
+                    filter,
+                    decorrelatedSubquery.getNode()));
+        }
+
+        checkState(
+                correlatedJoinNode.getType() == RIGHT || correlatedJoinNode.getType() == FULL,
```

Comment:
do we expect other kinds of joins in the future?

Reply:
I don't think we do, yet I think it should be explicit what types are handled where.

Reply:
@sopel39, this is similar to what needs to be done to ensure that all branches of a switch are exhaustive.